From: FIRERED project
Subject: [PATCH 6] Expand level-up learnsets to full 16-bit move IDs

Reference target: pret/pokefirered
Reference snapshot: c75f352304d529f6ba92d4f74b9cf8b5c3810788

Retail FireRed packs each level-up entry into one u16: 9 move bits + 7 level
bits. The expanded profile stores each logical entry as two u16 words:
move, level. Save data is unaffected.

--- a/include/pokemon.h
+++ b/include/pokemon.h
@@
-struct __attribute__((packed)) LevelUpMove
+struct LevelUpMove
 {
-    u16 move:9;
-    u16 level:7;
+    u16 move;
+    u16 level;
 };

--- a/include/constants/pokemon.h
+++ b/include/constants/pokemon.h
@@
-#define LEVEL_UP_MOVE_ID   0x01FF
-#define LEVEL_UP_MOVE_LV   0xFE00
 #define LEVEL_UP_END       0xFFFF
-
-#define MAX_LEVEL_UP_MOVES       20
+#define LEVEL_UP_MOVE_ENTRY_WORDS 2
+#define MAX_LEVEL_UP_MOVES       64

--- a/src/data/pokemon/level_up_learnsets.h
+++ b/src/data/pokemon/level_up_learnsets.h
@@
-#define LEVEL_UP_MOVE(lvl, move) ((lvl << 9) | move)
+#define LEVEL_UP_MOVE(lvl, move) (move), (lvl)
 #define LEVEL_UP_END 0xFFFF

--- a/src/pokemon.c
+++ b/src/pokemon.c
@@
-static EWRAM_DATA u8 sLearningMoveTableID = 0;
+static EWRAM_DATA u16 sLearningMoveTableID = 0;
@@
-u16 MonTryLearningNewMove(struct Pokemon *mon, bool8 firstMove)
-{
-    u32 retVal = MOVE_NONE;
-    u16 species = GetMonData(mon, MON_DATA_SPECIES, NULL);
-    u8 level = GetMonData(mon, MON_DATA_LEVEL, NULL);
-
-    // since you can learn more than one move per level
-    // the game needs to know whether you decided to
-    // learn it or keep the old set to avoid asking
-    // you to learn the same move over and over again
-    if (firstMove)
-    {
-        sLearningMoveTableID = 0;
-
-        while ((gLevelUpLearnsets[species][sLearningMoveTableID] & LEVEL_UP_MOVE_LV) != (level << 9))
-        {
-            sLearningMoveTableID++;
-            if (gLevelUpLearnsets[species][sLearningMoveTableID] == LEVEL_UP_END)
-                return MOVE_NONE;
-        }
-    }
-
-    if ((gLevelUpLearnsets[species][sLearningMoveTableID] & LEVEL_UP_MOVE_LV) == (level << 9))
-    {
-        gMoveToLearn = (gLevelUpLearnsets[species][sLearningMoveTableID] & LEVEL_UP_MOVE_ID);
-        sLearningMoveTableID++;
-        retVal = GiveMoveToMon(mon, gMoveToLearn);
-    }
-
-    return retVal;
-}
+u16 MonTryLearningNewMove(struct Pokemon *mon, bool8 firstMove)
+{
+    u32 retVal = MOVE_NONE;
+    u16 species = GetMonData(mon, MON_DATA_SPECIES, NULL);
+    u8 level = GetMonData(mon, MON_DATA_LEVEL, NULL);
+
+    if (firstMove)
+    {
+        sLearningMoveTableID = 0;
+
+        while (gLevelUpLearnsets[species][sLearningMoveTableID] != LEVEL_UP_END
+            && gLevelUpLearnsets[species][sLearningMoveTableID + 1] != level)
+            sLearningMoveTableID += LEVEL_UP_MOVE_ENTRY_WORDS;
+
+        if (gLevelUpLearnsets[species][sLearningMoveTableID] == LEVEL_UP_END)
+            return MOVE_NONE;
+    }
+
+    if (gLevelUpLearnsets[species][sLearningMoveTableID] != LEVEL_UP_END
+     && gLevelUpLearnsets[species][sLearningMoveTableID + 1] == level)
+    {
+        gMoveToLearn = gLevelUpLearnsets[species][sLearningMoveTableID];
+        sLearningMoveTableID += LEVEL_UP_MOVE_ENTRY_WORDS;
+        retVal = GiveMoveToMon(mon, gMoveToLearn);
+    }
+
+    return retVal;
+}
@@
-u8 GetMoveRelearnerMoves(struct Pokemon *mon, u16 *moves)
-{
-    u16 learnedMoves[MAX_MON_MOVES];
-    u8 numMoves = 0;
-    u16 species = GetMonData(mon, MON_DATA_SPECIES, NULL);
-    u8 level = GetMonData(mon, MON_DATA_LEVEL, NULL);
-    int i, j, k;
-
-    for (i = 0; i < MAX_MON_MOVES; i++)
-        learnedMoves[i] = GetMonData(mon, MON_DATA_MOVE1 + i, NULL);
-
-    for (i = 0; i < MAX_LEVEL_UP_MOVES; i++)
-    {
-        u16 moveLevel;
-
-        if (gLevelUpLearnsets[species][i] == LEVEL_UP_END)
-            break;
-
-        moveLevel = gLevelUpLearnsets[species][i] & LEVEL_UP_MOVE_LV;
-
-        if (moveLevel <= (level << 9))
-        {
-            for (j = 0; j < MAX_MON_MOVES && learnedMoves[j] != (gLevelUpLearnsets[species][i] & LEVEL_UP_MOVE_ID); j++)
-                ;
-
-            if (j == MAX_MON_MOVES)
-            {
-                for (k = 0; k < numMoves && moves[k] != (gLevelUpLearnsets[species][i] & LEVEL_UP_MOVE_ID); k++)
-                    ;
-
-                if (k == numMoves)
-                    moves[numMoves++] = gLevelUpLearnsets[species][i] & LEVEL_UP_MOVE_ID;
-            }
-        }
-    }
-
-    return numMoves;
-}
+u8 GetMoveRelearnerMoves(struct Pokemon *mon, u16 *moves)
+{
+    u16 learnedMoves[MAX_MON_MOVES];
+    u8 numMoves = 0;
+    u16 species = GetMonData(mon, MON_DATA_SPECIES, NULL);
+    u8 level = GetMonData(mon, MON_DATA_LEVEL, NULL);
+    int i, j, k;
+
+    for (i = 0; i < MAX_MON_MOVES; i++)
+        learnedMoves[i] = GetMonData(mon, MON_DATA_MOVE1 + i, NULL);
+
+    for (i = 0; i < MAX_LEVEL_UP_MOVES * LEVEL_UP_MOVE_ENTRY_WORDS; i += LEVEL_UP_MOVE_ENTRY_WORDS)
+    {
+        u16 move;
+        u16 moveLevel;
+
+        if (gLevelUpLearnsets[species][i] == LEVEL_UP_END)
+            break;
+
+        move = gLevelUpLearnsets[species][i];
+        moveLevel = gLevelUpLearnsets[species][i + 1];
+
+        if (moveLevel <= level)
+        {
+            for (j = 0; j < MAX_MON_MOVES && learnedMoves[j] != move; j++)
+                ;
+
+            if (j == MAX_MON_MOVES)
+            {
+                for (k = 0; k < numMoves && moves[k] != move; k++)
+                    ;
+
+                if (k == numMoves)
+                    moves[numMoves++] = move;
+            }
+        }
+    }
+
+    return numMoves;
+}
@@
-u8 GetLevelUpMovesBySpecies(u16 species, u16 *moves)
-{
-    u8 numMoves = 0;
-    int i;
-
-    for (i = 0; i < MAX_LEVEL_UP_MOVES && gLevelUpLearnsets[species][i] != LEVEL_UP_END; i++)
-         moves[numMoves++] = gLevelUpLearnsets[species][i] & LEVEL_UP_MOVE_ID;
-
-     return numMoves;
-}
+u8 GetLevelUpMovesBySpecies(u16 species, u16 *moves)
+{
+    u8 numMoves = 0;
+    int i;
+
+    for (i = 0;
+         i < MAX_LEVEL_UP_MOVES * LEVEL_UP_MOVE_ENTRY_WORDS
+         && gLevelUpLearnsets[species][i] != LEVEL_UP_END;
+         i += LEVEL_UP_MOVE_ENTRY_WORDS)
+        moves[numMoves++] = gLevelUpLearnsets[species][i];
+
+    return numMoves;
+}
@@
-u8 GetNumberOfRelearnableMoves(struct Pokemon *mon)
-{
-    u16 learnedMoves[MAX_MON_MOVES];
-    u16 moves[MAX_LEVEL_UP_MOVES];
-    u8 numMoves = 0;
-    u16 species = GetMonData(mon, MON_DATA_SPECIES_OR_EGG, NULL);
-    u8 level = GetMonData(mon, MON_DATA_LEVEL, NULL);
-    int i, j, k;
-
-    if (species == SPECIES_EGG)
-        return 0;
-
-    for (i = 0; i < MAX_MON_MOVES; i++)
-        learnedMoves[i] = GetMonData(mon, MON_DATA_MOVE1 + i, NULL);
-
-    for (i = 0; i < MAX_LEVEL_UP_MOVES; i++)
-    {
-        u16 moveLevel;
-
-        if (gLevelUpLearnsets[species][i] == LEVEL_UP_END)
-            break;
-
-        moveLevel = gLevelUpLearnsets[species][i] & LEVEL_UP_MOVE_LV;
-
-        if (moveLevel <= (level << 9))
-        {
-            for (j = 0; j < MAX_MON_MOVES && learnedMoves[j] != (gLevelUpLearnsets[species][i] & LEVEL_UP_MOVE_ID); j++)
-                ;
-
-            if (j == MAX_MON_MOVES)
-            {
-                for (k = 0; k < numMoves && moves[k] != (gLevelUpLearnsets[species][i] & LEVEL_UP_MOVE_ID); k++)
-                    ;
-
-                if (k == numMoves)
-                    moves[numMoves++] = gLevelUpLearnsets[species][i] & LEVEL_UP_MOVE_ID;
-            }
-        }
-    }
-
-    return numMoves;
-}
+u8 GetNumberOfRelearnableMoves(struct Pokemon *mon)
+{
+    u16 learnedMoves[MAX_MON_MOVES];
+    u16 moves[MAX_LEVEL_UP_MOVES];
+    u8 numMoves = 0;
+    u16 species = GetMonData(mon, MON_DATA_SPECIES_OR_EGG, NULL);
+    u8 level = GetMonData(mon, MON_DATA_LEVEL, NULL);
+    int i, j, k;
+
+    if (species == SPECIES_EGG)
+        return 0;
+
+    for (i = 0; i < MAX_MON_MOVES; i++)
+        learnedMoves[i] = GetMonData(mon, MON_DATA_MOVE1 + i, NULL);
+
+    for (i = 0; i < MAX_LEVEL_UP_MOVES * LEVEL_UP_MOVE_ENTRY_WORDS; i += LEVEL_UP_MOVE_ENTRY_WORDS)
+    {
+        u16 move;
+        u16 moveLevel;
+
+        if (gLevelUpLearnsets[species][i] == LEVEL_UP_END)
+            break;
+
+        move = gLevelUpLearnsets[species][i];
+        moveLevel = gLevelUpLearnsets[species][i + 1];
+
+        if (moveLevel <= level)
+        {
+            for (j = 0; j < MAX_MON_MOVES && learnedMoves[j] != move; j++)
+                ;
+
+            if (j == MAX_MON_MOVES)
+            {
+                for (k = 0; k < numMoves && moves[k] != move; k++)
+                    ;
+
+                if (k == numMoves)
+                    moves[numMoves++] = move;
+            }
+        }
+    }
+
+    return numMoves;
+}

@@
-static void GiveBoxMonInitialMoveset(struct BoxPokemon *boxMon)
-{
-    u16 species = GetBoxMonData(boxMon, MON_DATA_SPECIES, NULL);
-    s32 level = GetLevelFromBoxMonExp(boxMon);
-    s32 i;
-
-    for (i = 0; gLevelUpLearnsets[species][i] != LEVEL_UP_END; i++)
-    {
-        u16 moveLevel;
-        u16 move;
-
-        moveLevel = (gLevelUpLearnsets[species][i] & LEVEL_UP_MOVE_LV);
-
-        if (moveLevel > (level << 9))
-            break;
-
-        move = (gLevelUpLearnsets[species][i] & LEVEL_UP_MOVE_ID);
-
-        if (GiveMoveToBoxMon(boxMon, move) == MON_HAS_MAX_MOVES)
-            DeleteFirstMoveAndGiveMoveToBoxMon(boxMon, move);
-    }
-}
+static void GiveBoxMonInitialMoveset(struct BoxPokemon *boxMon)
+{
+    u16 species = GetBoxMonData(boxMon, MON_DATA_SPECIES, NULL);
+    s32 level = GetLevelFromBoxMonExp(boxMon);
+    s32 i;
+
+    for (i = 0; gLevelUpLearnsets[species][i] != LEVEL_UP_END; i += LEVEL_UP_MOVE_ENTRY_WORDS)
+    {
+        u16 move = gLevelUpLearnsets[species][i];
+        u16 moveLevel = gLevelUpLearnsets[species][i + 1];
+
+        if (moveLevel > level)
+            break;
+
+        if (GiveMoveToBoxMon(boxMon, move) == MON_HAS_MAX_MOVES)
+            DeleteFirstMoveAndGiveMoveToBoxMon(boxMon, move);
+    }
+}
